File: world_py/world/move.py
from redis import Redis

from structs import Move


def move(data: Move, redis: Redis):
    world_id: str = data.world_id

    world_state = redis.hget("world:"+world_id+":", "world_state")

    if world_state == "0":
        return

    x: int = int(data.agent_x)
    y: int = int(data.agent_y)

    if data.direction == "N":
        if y - 1 > 0:
            y = y - 1
    if data.direction == "S":
        # if y is within boundaries
        y = y + 1
    if data.direction == "W":
        if x - 1 > 0:
            x = x - 1
    if data.direction == "E":
        # if x is within boundaries
        x = x + 1

    # check if x:y is occupied
    target = redis.hget("world:"+world_id+":map", str(x)+":"+str(y))
    if target is not None:
        return

    # remove data from source node
    redis.hdel("world:"+world_id+":map", str(x)+":"+str(y))

    # add/update target node
    redis.hset("world:"+world_id+":map", "agent:"+data.agent_id, str(x)+":"+str(y))
    redis.hset("world:"+world_id+":map", str(x)+":"+str(y), "agent:"+data.agent_id)

    # update broadcast
    redis.hset("world:"+world_id+":path", "agent:"+data.agent_id, data.broadcast)

    # update Move Action Count
    redis.hincrby("world:"+world_id+":", "move_action_count", "1")

    return

File: world_go/handler/move.go
package handler

import (
	"github.com/gomodule/redigo/redis"
	"github.com/labstack/echo/v4"
	"net/http"
	"strconv"
	"strings"
)

type Move struct {
	Agent
	WorldID   string `json:"world_id" form:"world_id" query:"world_id"`
	Direction string `json:"direction" form:"direction" query:"direction"`
	Broadcast string `json:"broadcast" form:"broadcast" query:"broadcast"`
}

func (h *Handler) Move(ctx echo.Context) (err error) {
	move := new(Move)
	if err = ctx.Bind(move); err != nil {
		ctx.Logger().Fatal(err)
		return ctx.NoContent(http.StatusBadRequest)
	}

	agent := move.Agent
	x, err := strconv.Atoi(agent.X)
	y, err := strconv.Atoi(agent.Y)

	wid := move.WorldID

	_aid := "agent:" + agent.Id
	_wid := "world:" + wid + ":"
	_map := _wid + "map"
	_path := _wid + "path"
	_curr := agent.X + ":" + agent.Y

	dir := move.Direction
	dir = strings.ToUpper(dir)

	rds := h.Pool.Get()
	defer rds.Close()

	// check if move state
	worldState, err := redis.Int(rds.Do("HGET", _wid, "world_state"))
	if err != nil {
		ctx.Logger().Fatal(err)
	}
	if worldState == 0 {
		return ctx.NoContent(http.StatusForbidden)
	}

	switch dir {
	case "N":
		if y-1 >= 0 {
			y = y - 1
		}
	case "S":
		// process world size limit
		y = y + 1
	case "W":
		if x-1 >= 0 {
			x = x - 1
		}
	case "E":
		// process world size limit
		x = x + 1
	default:
	}

	xs := strconv.Itoa(x)
	ys := strconv.Itoa(y)

	_next := xs + ":" + ys

	// check if next x:y is occupied
	dest, _ := redis.String(rds.Do("HGET", _map, _next))
	if len(dest) > 0 {
		return ctx.NoContent(http.StatusForbidden)
	}

	// DEL world:{wid}:map x:y
	_, err = rds.Do("HDEL", _map, _curr)
	if err != nil {
		ctx.Logger().Fatal(err)
	}

	// update data
	agent.X = xs
	agent.Y = ys

	// update REDIS position
	_, err = rds.Do("HSET", _map, _aid, _next)
	_, err = rds.Do("HSET", _map, _next, _aid)
	if err != nil {
		ctx.Logger().Fatal(err)
	}

	// update REDIS move
	_, err = rds.Do("HSET", _path, _aid, move.Broadcast)
	if err != nil {
		ctx.Logger().Fatal(err)
	}

	// increment REDIS move action count
	_, err = rds.Do("HINCRBY", _wid, "move_action_count", "1")
	if err != nil {
		ctx.Logger().Error("Move: HINCRBY move_action_count 1 > ", err)
	}

	return ctx.JSON(http.StatusOK, agent)
}
